pkg/agent/controller/networkpolicy/l7engine/l7visbility_watcher.go

```diff
@@ -0,0 +1,364 @@
+		return
+	}
+	klog.V(2).InfoS("Processing Namespace UPDATE event", "Namespace", klog.KObj(ns))
+
+	affectedPods := l7w.getPodsFromNS(ns)
+	for _, pod := range affectedPods {
+		l7w.queue.Add(pod)
+	}
+}
+
+func (l7w *l7Watcher) deleteNamespace(old interface{}) {
+	namespace, ok := old.(*v1.Namespace)
+	if !ok {
+		tombstone, ok := old.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Error decoding object when deleting Namespace, invalid type: %v", old)
+			return
+		}
+		namespace, ok = tombstone.Obj.(*v1.Namespace)
+		if !ok {
+			klog.Errorf("Error decoding object tombstone when deleting Namespace, invalid type: %v", tombstone.Obj)
+			return
+		}
+	}
+	klog.V(2).Infof("Processing Namespace %s DELETE event, labels: %v", namespace.Name, namespace.Labels)
+	// l7w.groupEntityIndex.DeleteNamespace(namespace)
+
+	affectedPods := l7w.getPodsFromNS(namespace)
+	for _, pod := range affectedPods {
+		l7w.queue.Add(pod)
+	}
+}
+
+func (l7w *l7Watcher) getPodsFromNS(ns *v1.Namespace) []*v1.Pod {
+	var selectedPods []*v1.Pod
+	podSelector := labels.Everything()
+	pods, err := l7w.PodLister.Pods(ns.Name).List(podSelector)
+	if err != nil {
+		klog.ErrorS(err, "Could not list pods for given namespace", "namespace", ns)
+		return nil
+	}
+	selectedPods = append(selectedPods, pods...)
+	return selectedPods
+}
+
+func (l7w *l7Watcher) syncPodEvent(pod *v1.Pod) error {
+	if _, err := l7w.PodLister.Pods(pod.Namespace).Get(pod.Name); err != nil {
+		// Remove the TC flows
+		return l7w.removeTCFlow(pod)
+	}
+	annotationValue, ok := pod.Annotations[l7VisibilityAnnotationKey]
+	if !ok {
+		return l7w.removeTCFlow(pod)
+	}
+
+	// Check if the annotation value is one of the specified values
+	direction := v1alpha2.DirectionBoth
+	annotationValue = strings.ToLower(annotationValue)
+	switch annotationValue {
+	case "ingress":
+		direction = v1alpha2.DirectionIngress
+	case "egress":
+		direction = v1alpha2.DirectionEgress
+	case "both":
+		direction = v1alpha2.DirectionBoth
+	default:
+		return fmt.Errorf("annotation key %s can only have values (Ingress/Egress/Both)", l7VisibilityAnnotationKey)
+	}
+
+	podNN := k8s.NamespacedName(pod.Namespace, pod.Name)
+	tcName := fmt.Sprintf("tc-%s", podNN)
+	if !l7w.bindPodToTrafficControl(podNN, tcName) {
+		return fmt.Errorf("Could not bind TCName %s and pod %s", tcName, podNN)
+	}
+	newOfPorts := sets.New[int32]()
+	podInterfaces := l7w.interfaceStore.GetContainerInterfacesByPod(pod.Name, pod.Namespace)
+	if len(podInterfaces) == 0 {
+		return fmt.Errorf("Interfaces of Pod %v not found", klog.KObj(pod))
+	}
+	newOfPorts.Insert(podInterfaces[0].OFPort)
+	var ofPorts []uint32
+	for _, port := range sets.List(newOfPorts) {
+		ofPorts = append(ofPorts, uint32(port))
+	}
+	var targetPort uint32
+	if itf, ok := l7w.interfaceStore.GetInterfaceByName(config.L7NetworkPolicyTargetPortName); ok {
+		targetPort = uint32(itf.OFPort)
+	}
```

Comment:
It doesn't handle the notfound case, and it should be better to get the targetPort and save it as member variable before processing any Pod as it's constant.

Reply:
Can we not make a fair assumption that TargetPort will always be there given we take take of generating L7NetworkPolicyTargetPortName initially in the agent code. I will add it as constant.

Reply:
What if L7NetworkPolicy is not enabled?

Reply:
yes, that is a corner case, I will create the ports in the starting if l7Visibility is enabled